### PR TITLE
Peter/async mm speedups

### PR DESCRIPTION
**Pull request by probicheaux**
# Description

Small optimizations to async code and new README

## Type of change

Please delete options that are not relevant.

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] This change requires a documentation update

## How has this change been tested, please provide a testcase or example of how you tested the change?

Spun up on GCP and benchmarked

## Any specific deployment considerations

None

## Docs

-   [ ] Docs updated? What were the changes:

### CODE REVIEW

File: inference/enterprise/parallel/README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# Parallel Robofolow Inference server
+
+Introducing the highly concurrent implementation of Roboflow Inference Server! This version of the server accepts and processes requests asynchronously, running the web server, preprocessing, auto batching, inference, and post processing all in separate threads to increase server fps throughput. Separate requests to the same model will be batched on the fly as allowed by `$MAX_BATCH_SIZE`, and then response handling will occurr independently. Images are passed via python's SharedMemory module to maximize throughput.
```
Comment (yeldarby) [inline inference/enterprise/parallel/README.md:3]: We should add a note here about the enterprise licensing.

File: inference/models/yolov8/yolov8_object_detection.py [hunk 2]
```diff
@@ -38,10 +39,15 @@ def predict(self, img_in: np.ndarray, **kwargs) -> Tuple[np.ndarray]:
         Returns:
             Tuple[np.ndarray]: NumPy array representing the predictions, including boxes, confidence scores, and class confidence scores.
         """
+        t = perf_counter()
         predictions = self.onnx_session.run(None, {self.input_name: img_in})[0]
+        print(f"Took {perf_counter() - t} seconds to onnx infer")
```
Comment (yeldarby) [inline inference/models/yolov8/yolov8_object_detection.py:44]: We probably want to use a logger that won't always print this

File: inference/enterprise/parallel/parallel_http_config.py [hunk 3]
```diff
@@ -14,13 +14,7 @@
...
     raise RuntimeError("Redis must be configured to use async inference")
 pool = ConnectionPool(host=REDIS_HOST, port=REDIS_PORT, decode_responses=True)
-checker = ResultsChecker(
-    AsyncRedis(
-        host=REDIS_HOST, port=REDIS_PORT, decode_responses=True, max_connections=1000
-    )
-)
-model_manager = DispatchModelManager(model_registry, checker)
-model_manager.init_pingback()
+model_manager = None
```
Comment (probicheaux) [inline inference/enterprise/parallel/parallel_http_config.py:17]: Gross hack for now: initializing this stuff in the on app startup to make sure that model manager and web server are using same event loop

File: inference/enterprise/parallel/README.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
```
Comment (capjamesg) [inline inference/enterprise/parallel/README.md:3]: Please replace the first sentence with something like:

```
The Roboflow Inference Server supports concurrent processing.
```

We should be direct in the first sentence. It is the hook.

Learn more: https://jamesg.blog/2023/12/03/first-sentences/

Comment (capjamesg) [reply inference/enterprise/parallel/README.md:3]: > server fps throughput

FPS needs capitals.

Comment (capjamesg) [reply inference/enterprise/parallel/README.md:3]: > Images are passed via python's SharedMemory module to maximize throughput.

"python" needs a capital letter.

File: inference/enterprise/parallel/README.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+> ⚠️ Currently, only Object Detection, Instance Segmentation, and Classification models are supported by this module. Core models are not enabled
```
Comment (capjamesg) [inline inference/enterprise/parallel/README.md:5]: There is a missing period.

File: inference/enterprise/parallel/README.md (hunk 1 above)
Comment (probicheaux) [reply inference/enterprise/parallel/README.md:3]: moved it up in the readme

File: inference/enterprise/parallel/README.md [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+> ⚠️ We require a Roboflow Enterprise License to use this in production. See LICENSE.txt for details
+
+## How Do I use it?
```
Comment (capjamesg) [inline inference/enterprise/parallel/README.md:9]: Suggestion:

> ## How to Use Concurrent Processing

File: inference/enterprise/parallel/README.md [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+You can build the server using `./inference/enterprise/parallel/build.sh` and run it using `./inference/enterprise/parallel/run.sh`
+
+We provide a container at dockerhub that you can pull using `docker pull roboflow/roboflow-inference-server-gpu-parallel:latest`. If you are pulling a pinned tag, be sure to change the `$TAG` variable in `run.sh`.
```
Comment (capjamesg) [inline inference/enterprise/parallel/README.md:12]: `Docker Hub`.

File: inference/models/yolov8/yolov8_object_detection.py (hunk 2 above)
Comment (probicheaux) [reply inference/models/yolov8/yolov8_object_detection.py]: yeah, this was debug stuff -- removed

File: inference/enterprise/parallel/README.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+This is a drop in replacement for the old server, so you can send requests using the [same API calls](https://inference.roboflow.com/quickstart/http_inference/#step-2-run-inference) you were using previously
```
Comment (capjamesg) [inline inference/enterprise/parallel/README.md:14]: MIssing period.

File: inference/enterprise/parallel/README.md [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+
+## Performance
+We measure and report performance across a variety of different task types, by selecting random models found on Roboflow Universe.
```
Comment (capjamesg) [inline inference/enterprise/parallel/README.md:18]: No comma is needed here.

File: inference/enterprise/parallel/README.md [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+### Methodology
+
+The following metrics are taken on a machine with 8 cores and 1 T4 gpu. The fps metrics reflect best out of 3 trials.
```
Comment (capjamesg) [inline inference/enterprise/parallel/README.md:22]: "1" should be "one". "3" should be "three"

Numbers below 10 should be written out (i.e. "one", "two", "three").

**Review by yeldarby — APPROVED** [review]

**Review by yeldarby — APPROVED** [review]

**Review by paulguerrie — APPROVED** [review]

**Pull request merged**